addons/udes_permissions/models/user_template.py

```diff
@@ -0,0 +1,303 @@
+from odoo import models, fields, api
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class UserTemplate(models.Model):
+    """
+    TODO(palabaster): Ideally we'd just have a new tree view type which gives us a hierarchical view - look into this later
+
+    UserTemplate Model.
+
+    Allows users to configure a hierarchy of User Templates which have res.groups associated with them.
+    Parents inherit their descendants groups (and thus, ACLs) via non stored computes.
+
+    Has one magic field `template_group_id` which will automatically create/delete
+    a user group with the template which allows for easy tuning of ACLs for users using this template.
+    Name changes back / forth of this template should propagate between the group and template
+
+    If you are utilizing this model, you should only need to use the following functions;
+    - def get_template_groups()
+    - def get_template_descendants()
+    instead of directly access the fields on here.
+    """
+
+    _name = "user.template"
+    _description = "User Template"
+    _rec_name = "complete_name"
```

Comment:
Q: Is `rec_name` set to `complete_name` to help with the hierarchy view? If you have more than a few levels then this can get difficult to read pretty quickly.

Reply:
Yeah it's basically to make display_name based on the full hierarchy. Happy to change it back to just name if we would prefer

Reply:
we can do that for the tree view as we can just display a different column, but not the hierarchical tree view as that uses display_name in the widget

Reply:
Okay understood. Think there'll be some cases we do just want the name to display, but we can tackle that in the relevant future stories. :)